File: packages/wasm/as/assembly/msgpack/Write.ts
import { BigInt, BigNumber } from "../math";
import { Context } from "../debug";
import { JSON } from "../json";
import { Box } from "../containers";

export abstract class Write {
  abstract writeNil(): void;
  abstract writeBool(value: bool): void;
  abstract writeInt8(value: i8): void;
  abstract writeInt16(value: i16): void;
  abstract writeInt32(value: i32): void;
  abstract writeUInt8(value: u8): void;
  abstract writeUInt16(value: u16): void;
  abstract writeUInt32(value: u32): void;
  abstract writeFloat32(value: f32): void;
  abstract writeFloat64(value: f64): void;
  abstract writeStringLength(length: u32): void;
  abstract writeString(value: string): void;
  abstract writeBytesLength(length: u32): void;
  abstract writeBytes(value: ArrayBuffer): void;
  abstract writeBigInt(value: BigInt): void;
  abstract writeBigNumber(value: BigNumber): void;
  abstract writeJSON(value: JSON.Value): void;
  abstract writeArrayLength(length: u32): void;
  abstract writeArray<T>(
    a: Array<T>,
    fn: (writer: Write, item: T) => void
  ): void;
  abstract writeMapLength(length: u32): void;
  abstract writeMap<K, V>(
    m: Map<K, V>,
    key_fn: (writer: Write, key: K) => void,
    value_fn: (writer: Write, value: V) => void
  ): void;
  abstract writeExtGenericMap<K, V>(
    m: Map<K, V>,
    key_fn: (writer: Write, key: K) => void,
    value_fn: (writer: Write, value: V) => void
  ): void;

  abstract writeOptionalBool(value: Box<bool> | null): void;
  abstract writeOptionalInt8(value: Box<i8> | null): void;
  abstract writeOptionalInt16(value: Box<i16> | null): void;
  abstract writeOptionalInt32(value: Box<i32> | null): void;
  abstract writeOptionalUInt8(value: Box<u8> | null): void;
  abstract writeOptionalUInt16(value: Box<u16> | null): void;
  abstract writeOptionalUInt32(value: Box<u32> | null): void;
  abstract writeOptionalFloat32(value: Box<f32> | null): void;
  abstract writeOptionalFloat64(value: Box<f64> | null): void;
  abstract writeOptionalString(value: string | null): void;
  abstract writeOptionalBytes(value: ArrayBuffer | null): void;
  abstract writeOptionalBigInt(value: BigInt | null): void;
  abstract writeOptionalBigNumber(value: BigNumber | null): void;
  abstract writeOptionalJSON(value: JSON.Value | null): void;
  abstract writeOptionalArray<T>(
    a: Array<T> | null,
    fn: (writer: Write, item: T) => void
  ): void;
  abstract writeOptionalMap<K, V>(
    m: Map<K, V> | null,
    key_fn: (writer: Write, key: K) => void,
    value_fn: (writer: Write, value: V) => void
  ): void;
  abstract writeOptionalExtGenericMap<K, V>(
    m: Map<K, V> | null,
    key_fn: (writer: Write, key: K) => void,
    value_fn: (writer: Write, value: V) => void
  ): void;

  abstract context(): Context;
}

File: packages/wasm/rs/src/msgpack/write.rs
use super::error::EncodeError;
use crate::{BigInt, BigNumber, Context, JSON};
use core::hash::Hash;
use std::collections::BTreeMap;

pub trait Write {
    fn write_nil(&mut self) -> Result<(), EncodeError>;
    fn write_bool(&mut self, value: &bool) -> Result<(), EncodeError>;
    fn write_i8(&mut self, value: &i8) -> Result<(), EncodeError>;
    fn write_i16(&mut self, value: &i16) -> Result<(), EncodeError>;
    fn write_i32(&mut self, value: &i32) -> Result<(), EncodeError>;
    fn write_u8(&mut self, value: &u8) -> Result<(), EncodeError>;
    fn write_u16(&mut self, value: &u16) -> Result<(), EncodeError>;
    fn write_u32(&mut self, value: &u32) -> Result<(), EncodeError>;
    fn write_f32(&mut self, value: &f32) -> Result<(), EncodeError>;
    fn write_f64(&mut self, value: &f64) -> Result<(), EncodeError>;
    fn write_string_length(&mut self, length: &u32) -> Result<(), EncodeError>;
    fn write_string(&mut self, value: &str) -> Result<(), EncodeError>;
    fn write_bytes_length(&mut self, length: &u32) -> Result<(), EncodeError>;
    fn write_bytes(&mut self, buf: &[u8]) -> Result<(), EncodeError>;
    fn write_bigint(&mut self, value: &BigInt) -> Result<(), EncodeError>;
    fn write_bignumber(&mut self, value: &BigNumber) -> Result<(), EncodeError>;
    fn write_json(&mut self, value: &JSON::Value) -> Result<(), EncodeError>;
    fn write_array_length(&mut self, length: &u32) -> Result<(), EncodeError>;
    fn write_array<T: Clone>(
        &mut self,
        array: &[T],
        item_writer: impl FnMut(&mut Self, &T) -> Result<(), EncodeError>,
    ) -> Result<(), EncodeError>;
    fn write_map_length(&mut self, length: &u32) -> Result<(), EncodeError>;
    fn write_map<K, V: Clone>(
        &mut self,
        map: &BTreeMap<K, V>,
        key_writer: impl FnMut(&mut Self, &K) -> Result<(), EncodeError>,
        val_writer: impl FnMut(&mut Self, &V) -> Result<(), EncodeError>,
    ) -> Result<(), EncodeError>
    where
        K: Clone + Eq + Hash + Ord;
    fn write_ext_generic_map<K, V: Clone>(
        &mut self,
        map: &BTreeMap<K, V>,
        key_writer: impl FnMut(&mut Self, &K) -> Result<(), EncodeError>,
        val_writer: impl FnMut(&mut Self, &V) -> Result<(), EncodeError>,
    ) -> Result<(), EncodeError>
    where
        K: Clone + Eq + Hash + Ord;

    fn write_optional_bool(&mut self, value: &Option<bool>) -> Result<(), EncodeError>;
    fn write_optional_i8(&mut self, value: &Option<i8>) -> Result<(), EncodeError>;
    fn write_optional_i16(&mut self, value: &Option<i16>) -> Result<(), EncodeError>;
    fn write_optional_i32(&mut self, value: &Option<i32>) -> Result<(), EncodeError>;
    fn write_optional_u8(&mut self, value: &Option<u8>) -> Result<(), EncodeError>;
    fn write_optional_u16(&mut self, value: &Option<u16>) -> Result<(), EncodeError>;
    fn write_optional_u32(&mut self, value: &Option<u32>) -> Result<(), EncodeError>;
    fn write_optional_f32(&mut self, value: &Option<f32>) -> Result<(), EncodeError>;
    fn write_optional_f64(&mut self, value: &Option<f64>) -> Result<(), EncodeError>;
    fn write_optional_string(&mut self, value: &Option<String>) -> Result<(), EncodeError>;
    fn write_optional_bytes(&mut self, value: &Option<Vec<u8>>) -> Result<(), EncodeError>;
    fn write_optional_bigint(&mut self, value: &Option<BigInt>) -> Result<(), EncodeError>;
    fn write_optional_bignumber(&mut self, value: &Option<BigNumber>) -> Result<(), EncodeError>;
    fn write_optional_json(&mut self, value: &Option<JSON::Value>) -> Result<(), EncodeError>;
    fn write_optional_array<T: Clone>(
        &mut self,
        opt_array: &Option<Vec<T>>,
        item_writer: impl FnMut(&mut Self, &T) -> Result<(), EncodeError>,
    ) -> Result<(), EncodeError>;
    fn write_optional_map<K, V: Clone>(
        &mut self,
        opt_map: &Option<BTreeMap<K, V>>,
        key_writer: impl FnMut(&mut Self, &K) -> Result<(), EncodeError>,
        val_writer: impl FnMut(&mut Self, &V) -> Result<(), EncodeError>,
    ) -> Result<(), EncodeError>
    where
        K: Clone + Eq + Hash + Ord;
    fn write_optional_ext_generic_map<K, V: Clone>(
        &mut self,
        opt_map: &Option<BTreeMap<K, V>>,
        key_writer: impl FnMut(&mut Self, &K) -> Result<(), EncodeError>,
        val_writer: impl FnMut(&mut Self, &V) -> Result<(), EncodeError>,
    ) -> Result<(), EncodeError>
    where
        K: Clone + Eq + Hash + Ord;

    fn context(&mut self) -> &mut Context;
}
